### PR TITLE
refactors kubermatic-api main function

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:simply refactors `kubermatic-api` main function

Which issue(s) this PR fixes (optional, in fixes #<issue number>(, fixes #<issue_number>, ...) format, will close the issue(s) when PR gets merged):
See #2268 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
I'm working on https://app.zenhub.com/workspace/o/kubermatic/kubermatic/issues/2268 and noticed that I could refactor `main.go` func.

File: api/pkg/api/v1/types.go [hunk 1]
```diff
@@ -285,14 +290,15 @@ type Cluster struct {
...
+
 	// MachineNetworks optionally specifies the parameters for IPAM.
 	MachineNetworks []kubermaticv1.MachineNetworkingConfig `json:"machineNetworks,omitempty"`
 
 	// Version desired version of the kubernetes master components.
 	Version string `json:"version"`
 }
 
-// MarshalJSON marshals NewClusterSpec object into JSON. It is overwritten to control data
+// MarshalJSON marshals ClusterSpec object into JSON. It is overwritten to control data
```
Comment (maciaszczykm) [inline api/pkg/api/v1/types.go:301]: Thanks for that :)

File: api/cmd/kubermatic-api/main.go [hunk 2]
```diff
@@ -40,157 +39,151 @@ import (
...
+		}
+	}
 
-		clusterProviders[ctx] = kubernetesprovider.NewClusterProvider(
-			defaultImpersonationClientForSeed.CreateImpersonatedClientSet,
-			client.New(kubeInformerFactory.Core().V1().Secrets().Lister()),
-			kubermaticSeedInformerFactory.Kubermatic().V1().Clusters().Lister(),
-			workerName,
-		)
+	//
```
Comment (maciaszczykm) [inline api/cmd/kubermatic-api/main.go:124]: Remove this line?

File: api/cmd/kubermatic-api/main.go [hunk 3]
```diff
@@ -40,157 +39,151 @@ import (
...
+}
+
+func createInitProviders(options serverRunOptions) (providers, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", options.kubeconfig)
 	if err != nil {
-		glog.Fatal(err)
+		return providers{}, fmt.Errorf("unable to build client configuration from kubeconfig due to %v", err)
 	}
 
+	//
```
Comment (maciaszczykm) [inline api/cmd/kubermatic-api/main.go:82]: Remove this line?

File: api/cmd/kubermatic-api/main.go [hunk 4]
```diff
@@ -40,157 +39,151 @@ import (
...
-	projectMemberProvider := kubernetesprovider.NewProjectMemberProvider(defaultImpersonationClient.CreateImpersonatedClientSet, kubermaticMasterInformerFactory.Kubermatic().V1().UserProjectBindings().Lister())
 
-	// create a cluster provider for each context
-	clientcmdConfig, err := clientcmd.LoadFromFile(kubeconfig)
+	go metrics.ServeForever(options.internalAddr, "/metrics")
+	glog.Info(fmt.Sprintf("Listening on %s", options.listenAddress))
+	glog.Fatal(http.ListenAndServe(options.listenAddress, handlers.CombinedLoggingHandler(os.Stdout, apiHandler)))
+}
+
+func createInitProviders(options serverRunOptions) (providers, error) {
```
Comment (maciaszczykm) [inline api/cmd/kubermatic-api/main.go:76]: It is quite big, is there a way to split it a bit more?

**Review by maciaszczykm — COMMENTED** [review]
Looks good, I have few minor comments.

File: api/cmd/kubermatic-api/main.go (hunk 4 above)
Comment (maciaszczykm) [reply api/cmd/kubermatic-api/main.go:76]: It may make no sense as it is the initialization of everything. Then just forget about this comment :)

Comment (p0lyn0mial) [reply api/cmd/kubermatic-api/main.go:76]: ok :)

File: api/cmd/kubermatic-api/main.go (hunk 3 above)
Comment (p0lyn0mial) [reply api/cmd/kubermatic-api/main.go:82]: do you mind if I remove those lines in a follow up pull ?

Comment (maciaszczykm) [reply api/cmd/kubermatic-api/main.go:82]: No, no problem for me :)

**Review by maciaszczykm — APPROVED** [review]

**Pull request merged**